librz/il/vm_layer.c

```diff
@@ -0,0 +1,484 @@
+		lbl->addr = NULL;
+	}
+	rz_il_bv_free(lbl->addr);
+	free(lbl->label_id);
+	free(lbl);
+}
+
+static void free_opcode_kv(HtPPKv *kv) {
+	rz_il_bv_free(kv->key);
+	rz_pvector_free(kv->value);
+}
+
+static void free_bind_var_val(HtPPKv *kv) {
+	free(kv->key);
+}
+
+/**
+ * initiate an empty VM
+ * \param vm RzILVM, pointer to an empty VM
+ * \param start_addr ut64, initiation pc address
+ * \param addr_size int, size of the address in VM
+ * \param data_size int, size of the minimal data unit in VM
+ */
+RZ_API bool rz_il_vm_init(RzILVM vm, ut64 start_addr, int addr_size, int data_size) {
+	vm->addr_size = addr_size;
+	vm->data_size = data_size;
+
+	vm->vm_global_variable_list = (RzILVar *)calloc(VM_MAX_VAR, sizeof(RzILVar));
+	vm->vm_global_value_set = rz_il_new_bag(VM_MAX_VAL, (RzILBagFreeFunc)rz_il_free_value);
+
+	// Key : string
+	// Val : EffectLabel
+	// Do not dump it since its single signed here, and will be free in `close`
+	HtPPOptions lbl_options = { 0 };
+	lbl_options.cmp = (HtPPListComparator)strcmp;
+	lbl_options.hashfn = (HtPPHashFunction)sdb_hash;
+	lbl_options.dupkey = (HtPPDupKey)strdup;
+	lbl_options.dupvalue = NULL;
+	lbl_options.freefn = (HtPPKvFreeFunc)free_label_kv;
+	lbl_options.elem_size = sizeof(HtPPKv);
+	lbl_options.calcsizeK = (HtPPCalcSizeK)strlen;
+	vm->vm_global_label_table = ht_pp_new_opt(&lbl_options);
+
+	// Binding Table for Variable and Value
+	HtPPOptions bind_options = { 0 };
+	bind_options.cmp = (HtPPListComparator)strcmp;
+	bind_options.hashfn = (HtPPHashFunction)sdb_hash;
+	bind_options.dupkey = (HtPPDupKey)strdup;
+	bind_options.dupvalue = NULL;
+	bind_options.freefn = (HtPPKvFreeFunc)free_bind_var_val;
+	bind_options.elem_size = sizeof(HtPPKv);
+	bind_options.calcsizeK = (HtPPCalcSizeK)strlen;
+	vm->vm_global_bind_table = ht_pp_new_opt(&bind_options);
+
+	// Temporary Value for core theory execution
+	vm->temp_value_list = (RzILTemp *)calloc(VM_MAX_TEMP, sizeof(RzILTemp));
+	for (int i = 0; i < VM_MAX_TEMP; ++i) {
+		vm->temp_value_list[i] = rz_il_new_temp();
+	}
+
+	// TODO : More Arguments for vm init to control
+	//      1. Minimal unit size in memory
+	//      2. Multiple Memory
+	//      3. pc length
+	vm->mems = (Mem *)calloc(VM_MAX_TEMP, sizeof(Mem));
+	vm->pc = rz_il_bv_new_from_ut64(addr_size, start_addr);
+
+	// Table for storing the core theory opcodes
+	HtPPOptions ops_options = { 0 };
+	ops_options.cmp = (HtPPListComparator)rz_il_bv_cmp;
+	ops_options.hashfn = (HtPPHashFunction)rz_il_bv_hash;
+	ops_options.dupkey = (HtPPDupKey)rz_il_bv_dump;
+	ops_options.dupvalue = NULL; // dump key only, since the opcode used in hash map only
+	ops_options.freefn = free_opcode_kv;
+	ops_options.elem_size = sizeof(HtPPKv);
+	vm->ct_opcodes = ht_pp_new_opt(&ops_options);
+
+	// init jump table of labels
+	vm->op_handler_table = (RzILOpHandler *)malloc(sizeof(RzILOpHandler) * RZIL_OP_INVALID);
+	memset(vm->op_handler_table, 0, RZIL_OP_INVALID);
```

Comment:
Add special last item (for example `_RZIL_OP_NR_MAX`) to `CoreTheoryOPCode` enum and use it here instead of `RZIL_OP_INVALID`. Since `CoreTheoryOPCode` starts from 0 you can't use `vm->op_handler_table[RZIL_OP_INVALID]`.